### DISCUSSION TITLE
sequential or parallel cache mode configuration support

### DISCUSSION THREAD

**Discussion by YX1AOL** [ChampSim/ChampSim#553] [Q&A]
Hi everyone, I'm currently studying the performance of cache replacement policy on caches in sequential or parallel mode.I observed that GEM5 is handled like below code:

```cpp
Cycles
BaseCache::calculateAccessLatency(const CacheBlk* blk, const uint32_t delay,
                                  const Cycles lookup_lat) const
{
    Cycles lat(0);

    if (blk != nullptr) {
        // As soon as the access arrives, for sequential accesses first access
        // tags, then the data entry. In the case of parallel accesses the
        // latency is dictated by the slowest of tag and data latencies.
        if (sequentialAccess) {
            lat = ticksToCycles(delay) + lookup_lat + dataLatency;
        } else {
            lat = ticksToCycles(delay) + std::max(lookup_lat, dataLatency);
        }

        // Check if the block to be accessed is available. If not, apply the
        // access latency on top of when the block is ready to be accessed.
        const Tick tick = curTick() + delay;
        const Tick when_ready = blk->getWhenReady();
        if (when_ready > tick &&
            ticksToCycles(when_ready - tick) > lat) {
            lat += ticksToCycles(when_ready - tick);
        }
    } else {
        // In case of a miss, we neglect the data access in a parallel
        // configuration (i.e., the data access will be stopped as soon as
        // we find out it is a miss), and use the tag-only latency.
        lat = calculateTagOnlyLatency(delay, lookup_lat);
    }

    return lat;
}

```

It seems that Champsim has not supported the configuration of this feature, nor does it distinguish between tag latency and data latency during Cache access. I'm wondering what part of the code I should look at if I wanted to start adding this feature.

I also sincerely ask how the current version of Champsim handles Cache Latency. When I read the old version of the code (IPC1 version), I noticed the code for Latency processing of PQ, RQ, WQ, and MSHR, but I did not find a similar code in the new version. I hope someone can answer my doubts. Finally, thank you to all developers who have contributed to Champsim!

**Comment by ngober** [comment]
This is a very good and insightful question, and it makes me happy to answer it.

You are correct that ChampSim does not make the same distinction as gem5 between the tag latency and the data latency. It is all rolled into one single ["hit latency"](https://github.com/ChampSim/ChampSim/blob/ff13e761366801ed5d520d6745ea1776eaac825f/src/cache.cc#L394). This is one of two cache latencies that ChampSim models; the other is the [fill latency](https://github.com/ChampSim/ChampSim/blob/ff13e761366801ed5d520d6745ea1776eaac825f/src/cache.cc#L383). If someone wanted to model the sequential/parallel access, the easiest way would be to pre-calculate the sum or max that you describe, and configure ChampSim with the result. 

If you wanted to start working on a feature like this, the linked lines are probably near to where you would need to start.

**Reply by YX1AOL** [reply]
Thank you for your quick answer and thorough explanation,I will start researching how to implement this feature, and if there are any questions or progress I will continue to share them here.
